toolkit/tools/scheduler/schedulerutils/graphbuildstate.go

```diff
@@ -17,31 +19,50 @@ type nodeState struct {
 	available bool
 	cached    bool
 	usedDelta bool
+	freshness int
 }
 
+const (
+	// The highest possible freshness value, used to force unbounded cascading rebuilds.
+	NodeFreshnessAbsoluteMax int = math.MaxInt
+)
+
 // GraphBuildState represents the build state of a graph.
 type GraphBuildState struct {
 	activeBuilds     map[int64]*BuildRequest
 	nodeToState      map[*pkggraph.PkgNode]*nodeState
+	maxFreshness     int
 	failures         []*BuildResult
 	reservedFiles    map[string]bool
 	conflictingRPMs  map[string]bool
 	conflictingSRPMs map[string]bool
 }
 
 // NewGraphBuildState returns a new GraphBuildState.
-// - reservedFiles is a list of reserved files which should NOT be rebuilt. Any files that ARE rebuilt will be recorded.
-func NewGraphBuildState(reservedFiles []string) (g *GraphBuildState) {
+//   - reservedFiles is a list of reserved files which should NOT be rebuilt. Any files that ARE rebuilt will be recorded.
+//   - maxFreshness is how fresh a newly rebuilt node is. Each dependant node will have a freshness of 'n-1', etc. until
+//     '0' where the subsequent nodes will no longer be rebuilt. 'maxFreshness < 0' will cause unbounded cascading rebuilds,
+//     while 'maxFreshness = 0' will cause no cascading rebuilds.
+func NewGraphBuildState(reservedFiles []string, maxFreshness int) (g *GraphBuildState) {
 	filesMap := make(map[string]bool)
 	for _, file := range reservedFiles {
 		filesMap[file] = true
 	}
+
+	// If we select negative maxFreshness this means we want unbounded cascading. Use max int to represent infinity.
+	// (2^31 or 2^63 packages is much more than the size of our graph, so this has the desired effect)
+	if maxFreshness < 0 {
```

Comment:
What do you think about making `maxFreshness` a `uint` (including the argument for the scheduler and switching the maximum to `math.MaxUint`) to avoid special-casing negative numbers?

Reply:
Sure, sounds good. It was needed in the older version before you refactored the request building code, but now we can fully calculate the freshness before we even try to build so there is no need to send a special `-1` signal back on a rebuild from the worker. Will update alongside the new default value approach.